api/services/diff/v1/diff.proto

```diff
@@ -20,11 +20,17 @@ service Diff {
 	rpc Diff(DiffRequest) returns (DiffResponse);
 }
 
+message ByteArrays {
+	repeated bytes byte_arrays = 1;
+}
+
 message ApplyRequest {
 	// Diff is the descriptor of the diff to be extracted
 	containerd.types.Descriptor diff = 1;
 
 	repeated containerd.types.Mount mounts = 2;
+
+	map<string, ByteArrays> dcparameters = 3;
```

Comment:
How are these expected to differ than the annotations in the descriptor?

Reply:
This map holds the private keys (sorted by type) needed for decryption,

Reply:
dcparameter = 'decryption parameters'

Reply:
I guess I mean from a data modeling perspective. The private keys are mapped by key id right? Also they are related only to the descriptor, as it is the descriptor which is being decrypted.

My thought is why the decryption parameters aren't passed through in the descriptor annotations. The API is designed to be generic but this parameter seems very feature specific.

Reply:
The keys are passed in their raw format , so either in der or pem format for example. Once they are in this map, they have been sorted by their format. Ideally only a single key is found in this map to decrypt an image, but multiple keys can be passed as well. If an image is encrypted for multiple recipients in JWE and PKCS7 format then we will try to decrypt the layers by invoking the handlers for these encryption technologies until a decryption succeeds. The JWE decryptor will look up the keys it can use from the map and try one after another, if multiple are given.

Reply:
The descriptor annotations are carrying the encrypted material and are read from the image metadata documents. I am not sure why we would want to mix those with parameters passed in by the user for decryption.

Reply:
I agree that we should not change the interface if possible - i ran into some trouble with this myself when trying to integrate this with buildkit, and I was looking at passing it via the context, but I think annotations are also fine.

@stefanberger i think we can have the same map structure and usage of `dcparameters` as long as we properly namespace the annotations to match our existing map structure. The overhead would be the base64 encoding of the keys. i.e. `org.opencontainers...enc.dcparameters.[openpgp|jwe|...]`

>The descriptor annotations are carrying the encrypted material and are read from the image metadata documents. I am not sure why we would want to mix those with parameters passed in by the user for decryption.

I see what you mean here, and I think the separation between image and caller parameters details is nice to have, but I do think that adding a specific data structure for a specific feature to the interface doesn't really follow the containerd generic and pluggable model.  

Maybe one potential way to go is to have a generic option(s) parameter. But an interface breaking change like what probably may need to be planned for further ahead of time for other consumers. In the meantime, I think annotations provides a middle ground for us to work from.

Reply:
@lumjjb @dmcgowan  So are you suggesting we merge the keys into the annotations map and then unmerge them once received on the server side of the grpc? There are other places as well where we have an image descriptor read from image metadata from disk and the keys coming in from different sources (ctr or API) and then what ? Do we merge them as well so that now all functions pick out the keys from the annotations map? I don't think this is a nice solution with the Descriptor corresponding to the image metadata file one-to-o ne but then we find some keys in there as well. Besides that the keys are in binary format and would quite frequently have to be base64 decoded.
If the API compatibility is the issue I would merge them into the Descriptor for the grpc call but then take them out on the receiver side so that we have a distinct `dcparameters` and `ocispec.Descriptor` where one knows what the data structure carry.

Reply:
https://github.com/containerd/containerd/pull/2532/commits/8f7b1fc3761a69372684eae78bd2220221b020ed now reverts the changes to the Apply API call and we are funneling the private keys through the ocispec.Descriptor.Annotations via a key "_dcparameters" that we put in the on the server side (JSON representation of the sorted decryption key material) and taken out of the Annotations map once received on the server side. From then on the key material is again carried by a parameter called 'dcparameters.'